yksgi/src/lib.rs

```diff
@@ -425,38 +425,48 @@ impl SGInterp {
     unsafe fn gep(&mut self) {
         // FIXME: If the target is a `struct` and it's packed, then we need to get the data layout
         // to compute the correct offsets.
+        // FIXME: Look into inbounds keyword and overflows.
         let ty = self.pc.get_type();
         debug_assert!(ty.is_pointer());
 
         let layout = self.module.datalayout();
         let aggr = self.pc.get_operand(0);
         let mut curty = aggr.get_type();
-        let ptr = self.var_lookup(&aggr).val;
+        let ptr = self.var_lookup(&aggr).val as *const c_void;
         let numops = LLVMGetNumOperands(self.pc.get());
 
-        let mut offset = 0;
+        let mut offset: i64 = 0;
         for i in 1..numops {
             let op = self.pc.get_operand(u32::try_from(i).unwrap());
-            let idx = if op.is_constant() {
-                LLVMConstIntGetZExtValue(op.get())
-            } else {
-                self.var_lookup(&op).val
-            };
 
             if curty.is_struct() {
+                // Struct indexes are always positive.
+                let idx = if op.is_constant() {
+                    LLVMConstIntGetZExtValue(op.get())
+                } else {
+                    self.var_lookup(&op).val
+                };
                 let off = LLVMOffsetOfElement(layout, curty.get(), u32::try_from(idx).unwrap());
-                offset += off;
+                // Since vector indexes expect an i64, and struct indexes are limited to 32bit, we
+                // can savely cast to i64 here.
+                offset += off as i64;
```

Comment:
I think we can use `i64::from(off)` here (which is safer than `as`).

Reply:
It would have to be a `try_from` since `off` is ` u64`. Fixed in 837ac91.

Reply:
Ah, then `try_from` is the right thing, but the comment makes no sense to me as it mention "32 bit" -- but there doesn't seem to be 32 bits anywhere in this calculation?

Reply:
I've just revised the comment in ee2a5c8, but now realise, that this was referring to the index `idx`, which will always be i32 when indexing into a struct. But I am forcing the offset to be i64, which is not the same. For example if `idx` is `u32::max`, then `LLVMOffsetOfElement` will certainly be bigger than `u32::max` and thus not fit into `i64`, right? Which is the same worry I mentioned in my disclaimer above.

Reply:
Que?

Reply:
Forget what I said earlier. Try with [53d58d7](https://github.com/ykjit/yk/pull/552/commits/53d58d79bd9684e29eb6aa9ade86cdb4262941b8)